infrastructure-formula/infrastructure/libvirt/domains.sls

```diff
@@ -49,24 +50,103 @@ along with this program.  If not, see <https://www.gnu.org/licenses/>.
 {%- for machine, config in machinepillar.items() %}
 {%- set machine = machine ~ '.' ~ domain %}
 {%- if config['cluster'] == cluster %}
+{%- set domainxml = domaindir ~ '/' ~ machine ~ '.xml' %}
+{%- if salt['cmd.retcode']('grep -q uuid ' ~ domainxml) == 0 %}
+{%- set uuid = salt['cmd.run']('grep -oP "(?<=<uuid>).*(?=</uuid>)" ' ~ domainxml) %}
+{%- do salt.log.debug('infrastructure.libvirt: found existing uuid') %}
+{%- else %}
+{%- set uuid = salt['cmd.run']('uuidgen') %}
```

Comment:
The uuid handling is not really nice. Unfortunately I don't have a perfect solution for it :-/

An idea to make it slightly less ugly would be:

```
{%- if salt['file.file_exists'](domainxml) %}
{%- set uuid = salt['cmd.run']('grep -oP "(?<=<uuid>).*(?=</uuid>)" ' ~ domainxml ~ ' || uuidgen') %}
```

This avoids one of the grep calls, and also prevents a funny bug in your current code - what happens if you have a VM called `douuidstuff` (which means the first grep will succeed, even if it matches the wrong line), but doesn't have an uuid defined in domainxml?

Actually you could even drop the if condition and do the `grep ... || uuidgen` call for all cases - if the file doesn't exist, grep will fail, and uuidgen will do its job. But if you do this, you'll need to suppress the grep stderr ("file not found") - not really nice. Oh, and you'd loose the "found uuid//generated new uuid" log messages ;-)

Reply:
Can I not simply adjust the first grep call to search for a fixed `<uuid>` string?

Reply:
That's also an option, but I'm quite sure that it will be slower (an additional fork per VM) than using `file_exists` with the `|| uuidgen` fallback. (If the file exists, but grep "just" doesn't find a match, there isn't an error message, so there shouldn't be a problem.)

The "Actually ..." part of my idea indeed has issues with triggering an error message. I'm a bit surprised that stderr ends up in the variable though. (`2>/dev/null` might be a workaround, but will also hide other errors.)

Reply:
Ok, and good idea - what about the new variant?

Reply:
Looks good :-)

Reply:
Sweet, thanks. :)